### PR TITLE
TypePredicate - Parse the asserts modifier

### PR DESCRIPTION
**Pull request by dsherret**
Start of adding support for the asserts modifier:

```ts
function isString(x: unknown): asserts x is string {
    if (typeof x !== "string")
        throw new Error("Not a string.");
}
```

Start of adding the asserts modifier, but this doesn't compile.

More info: https://devblogs.microsoft.com/typescript/announcing-typescript-3-7/#assertion-functions

### CODE REVIEW

File: ecmascript/parser/src/parser/typescript.rs [hunk 1]
```diff
@@ -256,11 +256,15 @@ impl<'a, I: Tokens> Parser<'a, I> {
...
+    fn parse_ts_this_type_predicate(
+        &mut self,
+        start: BytePos,
+        has_asserts_keyword: bool,
+        lhs: TsThisType,
+    ) -> PResult<'a, TsTypePredicate> {
         debug_assert!(self.input.syntax().typescript());
 
         assert_and_bump!("is");
-        let start = cur_pos!();
```
Comment (dsherret) [inline ecmascript/parser/src/parser/typescript.rs]: Also, will fix this span issue.

File: ecmascript/parser/src/parser/typescript.rs [hunk 2]
```diff
@@ -1687,11 +1698,13 @@ impl<'a, I: Tokens> Parser<'a, I> {
                 return self.parse_ts_import_type().map(TsType::from).map(Box::new);
             }
 
-            tok!("this") => {
+            tok!("this") | tok!("asserts") if is!("this") || peek_is!("this") => {
```
Comment (dsherret) [inline ecmascript/parser/src/parser/typescript.rs]: How would you recommend checking for "asserts" here and above for `is!("asserts")`? This doesn't compile.

File: ecmascript/ast/src/typescript.rs [hunk 3]
```diff
@@ -397,6 +397,7 @@ pub struct TsTypeRef {
 #[ast_node("TsTypePredicate")]
 pub struct TsTypePredicate {
     pub span: Span,
+    pub asserts: bool,
```
Comment (dsherret) [inline ecmascript/ast/src/typescript.rs]: Is doing this boolean flag ok?

File: ecmascript/parser/src/parser/typescript.rs (hunk 1 above)
Comment (kdy1) [reply ecmascript/parser/src/parser/typescript.rs]: Great

File: ecmascript/parser/src/parser/typescript.rs (hunk 2 above)
Comment (kdy1) [reply ecmascript/parser/src/parser/typescript.rs]: You need to add "asserts" to tok! macro.

File: ecmascript/ast/src/typescript.rs (hunk 3 above)
Comment (kdy1) [reply ecmascript/ast/src/typescript.rs]: Yes. I think it's enough.

File: ecmascript/parser/src/parser/typescript.rs [hunk 4]
```diff
@@ -400,14 +405,19 @@ impl<'a, I: Tokens> Parser<'a, I> {
         debug_assert!(self.input.syntax().typescript());
 
         self.in_type().parse_with(|p| {
-            let start = cur_pos!();
+            let start = cur_pos!(); // todo: the starts below should not include the return_token
```
Comment (dsherret) [inline ecmascript/parser/src/parser/typescript.rs]: I believe everything is good in this PR, but it shouldn't use this start for type predicates in functions. I believe this should go after the if statement below, but might be good to fix that in a separate PR.

File: ecmascript/parser/tests/typescript/function/predicate-types/input.ts.json [hunk 5]
```diff
@@ -222,6 +224,113 @@
...
+        "type": "TsTypeAnnotation",
+        "span": {
+          "start": 90,
+          "end": 112,
+          "ctxt": 0
+        },
+        "typeAnnotation": {
+          "type": "TsTypePredicate",
+          "span": {
+            "start": 90,
```
Comment (dsherret) [inline ecmascript/parser/tests/typescript/function/predicate-types/input.ts.json]: As pointed out above, this should be 92, but fixing it causes too many other changes for this PR.

File: ecmascript/parser/src/parser/typescript.rs (hunk 4 above)
Comment (kdy1) [reply ecmascript/parser/src/parser/typescript.rs]: I agree with you.

**Comment by kdy1** [comment]
Thank you for the pr!

bors r+

**Comment by dsherret** [comment]
@kdy1 I'm not sure what's going on with the release build test failing so if you have any ideas feel free to push a fix.
